### PR TITLE
Add openshift mgr logic directly to pman as an option

### PR DESCRIPTION
**Pull request by danmcp**
OpenShift is being added here as a sibling to "container".  I looked into adding the logic under the _container logic and it didn't go very smoothly.  In the end I came to the realization that "container" really means it's going to interact with the docker API and therefore a sibling element was the right level to abstract everything out into a different code path.

### CODE REVIEW

**Comment by danmcp** [comment]
@rudolphpienaar Let me know what you think.

**Comment by danmcp** [comment]
@awalkaradi95moc FYI

File: pman/openshiftmgr.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
+"""
+OpenShift cluster manager module that provides functionality to schedule jobs as well as
+manage their state in the cluster.
+"""
+
+import yaml
+import json
+import os
+from kubernetes import client
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: can this be renamed to kube_clientm just to avoid confusion?

File: pman/openshiftmgr.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+    def init_openshift_client(self):
+        """
+        Method to get a OpenShift client connected to remote or local OpenShift.
+        """
+        kubecfg_path = os.environ.get('KUBECFG_PATH')
+        if kubecfg_path is None:
+            config.load_kube_config()
+        else:
+            config.load_kube_config(config_file=kubecfg_path)
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: We need to have a check here to see if kubecfg_path exists or is load_kube_config capable of that?

File: pman/openshiftmgr.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        self.openshift_client = o_client.OapiApi()
+        self.kube_client = client.CoreV1Api()
+        self.kube_v1_batch_client = client.BatchV1Api()
+
+    def schedule(self, image, command, name, project=None, mountdir=None):
+        """
+        Schedule a new job and returns the job object.
+        """
+        project = project or self.project
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: Any specific reason for having a project parameter? When we are instantiating this class, we can ask project to be passed, if None, we won't proceed? IOW, why can't we have just self.project?

File: pman/openshiftmgr.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        job_str = """
+apiVersion: batch/v1
+kind: Job
+metadata:
+    name: {name}
+spec:
+    parallelism: 1
+    completions: 1
+    activeDeadlineSeconds: 3600
+    template:
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: Any specific reason for having a string? Why can't we load this as yaml file? My only concern is if we want to have parameterize parallelism, completions etc, I believe we have to change the code.

File: pman/openshiftmgr.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+              - mountPath: /share
+                name: openshiftmgr-storage
+            volumes: 
+            - name: openshiftmgr-storage
+              hostPath:
+                path: {mountdir}
+""".format(mountdir=mountdir)
+
+        job_yaml = yaml.load(job_str)
+        job = self.kube_v1_batch_client.create_namespaced_job(namespace=project, body=job_yaml)
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: create_namespaced_job seems to be async call, should we have a callback to check if it created successfully or not?

File: pman/openshiftmgr.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        log = self.kube_client.read_namespaced_pod_status(namespace=project, name=name)
+        return log
+
+    def get_pod_log(self, name, project=None):
+        """
+        Get a pod log
+        """
+        project = project or self.project
+        log = self.kube_client.read_namespaced_pod_log(namespace=project, name=name)
+        return log
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: Ditto as above method.

File: pman/openshiftmgr.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+    def state(self, name, project=None):
+        """
+        Return the state of a previously scheduled job.
+        """
+        project = project or self.project
+        job = self.get_job(name, project)
+        message = None
+        state = None
+        reason = None
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: nit: all of them could have been empty strings, just for sake of readability.

File: pman/openshiftmgr.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                state = 'inactive'
+
+        return {'Status': {'Message': message,
+                                'State': state,
+                                'Reason': reason,
+                                'Active': job.status.active,
+                                'Failed': job.status.failed,
+                                'Succeeded': job.status.succeeded,
+                                'StartTime': job.status.start_time,
+                                'CompletionTime': job.status.completion_time}}
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: nit: empty line

File: pman/openshiftmgr.py (hunk 2 above)
Comment (danmcp) [inline pman/openshiftmgr.py]: It will error if the path is invalid.  Which I think is desired.

**Comment by danmcp** [comment]
@rudolphpienaar

**Pull request merged**